kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ProtocolMetadataLogReplay.java

```diff
@@ -0,0 +1,244 @@
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.lang.Lazy;
+import io.delta.kernel.internal.metrics.SnapshotMetrics;
+import io.delta.kernel.internal.snapshot.LogSegment;
+import io.delta.kernel.internal.tablefeatures.TableFeatures;
+import io.delta.kernel.internal.util.FileNames;
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.FileStatus;
+import java.io.IOException;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Static utility class for loading Protocol and Metadata from Delta log files. */
+public class ProtocolMetadataLogReplay {
+
+  private static final Logger logger = LoggerFactory.getLogger(ProtocolMetadataLogReplay.class);
+
+  /** Read schema when searching for the latest Protocol and Metadata. */
+  public static final StructType PROTOCOL_METADATA_READ_SCHEMA =
+      new StructType().add("protocol", Protocol.FULL_SCHEMA).add("metaData", Metadata.FULL_SCHEMA);
+
+  /** Result of loading Protocol and Metadata from a LogSegment. */
+  public static class Result {
+    public final Protocol protocol;
+    public final Metadata metadata;
+    public final CachedCrcInfoResult crcInfoUsed;
+    private final long logFilesRead;
+
+    public Result(
+        Protocol protocol, Metadata metadata, CachedCrcInfoResult crcInfoUsed, long logFilesRead) {
+      this.protocol = protocol;
+      this.metadata = metadata;
+      this.crcInfoUsed = crcInfoUsed;
+      this.logFilesRead = logFilesRead;
+    }
+  }
+
+  /**
+   * Loads the latest Protocol and Metadata from the log files in the given LogSegment.
+   *
+   * <p>Uses the {@code LogSegment::lastSeenChecksum} to bound how many delta files it reads. That
+   * is, we only need to read delta files *newer* than that CRC file to search for any new P & M. If
+   * we don't find both the P & M by the time we get to the CRC file version, we can just use the P
+   * and M from it.
+   *
+   * <p>Also validates that the Kernel can read the table at the loaded protocol version.
+   */
+  public static Result loadProtocolAndMetadata(
+      Engine engine, LogSegment logSegment, Path dataPath, SnapshotMetrics snapshotMetrics) {
+
+    final Result result =
+        snapshotMetrics.loadProtocolMetadataTotalDurationTimer.time(
+            () -> loadProtocolAndMetadataInternal(engine, logSegment, dataPath, snapshotMetrics));
+
+    TableFeatures.validateKernelCanReadTheTable(result.protocol, dataPath.toString());
+
+    logger.info(
+        "[{}] Took {}ms to load Protocol and Metadata at version {}. {}, read {} log files",
+        dataPath,
+        snapshotMetrics.loadProtocolMetadataTotalDurationTimer.totalDurationMs(),
+        logSegment.getVersion(),
+        buildCrcStatusMessage(logSegment, result.crcInfoUsed),
+        result.logFilesRead);
+
+    return result;
+  }
+
+  private static Result loadProtocolAndMetadataInternal(
```

Comment:
Mostly moved verbatim from `LogReplay.java`. Changes:
- crc loading is truly lazy now
- we return P & M & the cached CRC info read result
- clearer logging done in one place only